allennlp_models/rc/dataset_readers/squad2.py

```diff
@@ -0,0 +1,197 @@
+        file_path = cached_path(file_path)
+
+        logger.info("Reading file at %s", file_path)
+        with open(file_path) as dataset_file:
+            dataset_json = json.load(dataset_file)
+            dataset = dataset_json["data"]
+        logger.info("Reading the dataset")
+        for article in dataset:
+            for paragraph_json in article["paragraphs"]:
+                paragraph = paragraph_json["context"]
+                tokenized_paragraph = self._tokenizer.tokenize(paragraph)
+                for question_answer in paragraph_json["qas"]:
+                    question_text = question_answer["question"].strip().replace("\n", "")
+                    is_impossible = question_answer["is_impossible"]
+                    if is_impossible:
+                        answer_texts: List[str] = []
+                        span_starts: List[int] = []
+                        span_ends: List[int] = []
+                    else:
+                        answer_texts = [answer["text"] for answer in question_answer["answers"]]
+                        span_starts = [
+                            answer["answer_start"] for answer in question_answer["answers"]
+                        ]
+                        span_ends = [
+                            start + len(answer) for start, answer in zip(span_starts, answer_texts)
+                        ]
+                    additional_metadata = {"id": question_answer.get("id", None)}
+                    instance = self.text_to_instance(
+                        question_text,
+                        paragraph,
+                        is_impossible,
+                        zip(span_starts, span_ends),
+                        answer_texts,
+                        tokenized_paragraph,
+                        additional_metadata,
+                    )
+                    if instance is not None:
+                        yield instance
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        question_text: str,
+        passage_text: str,
+        is_impossible: bool = None,
+        char_spans: List[Tuple[int, int]] = None,
+        answer_texts: List[str] = None,
+        passage_tokens: List[Token] = None,
+        additional_metadata: Dict[str, Any] = None,
+    ) -> Optional[Instance]:
+
+        if not passage_tokens:
+            passage_tokens = self._tokenizer.tokenize(passage_text)
+        question_tokens = self._tokenizer.tokenize(question_text)
+        if self.passage_length_limit is not None:
+            passage_tokens = passage_tokens[: self.passage_length_limit]  # type: ignore
+
+        passage_tokens_with_no_answer = passage_tokens + [  # type: ignore
+            Token(
+                text=SQUAD2_NO_ANSWER_TOKEN,
+                idx=passage_tokens[-1].idx + len(passage_tokens[-1].text) + 1,  # type: ignore
+                lemma_=SQUAD2_NO_ANSWER_TOKEN,
+            )
+        ]
```

Comment:
Now `passage_tokens_with_no_answer` will be one token too long, won't it?

Reply:
I chose to interpret the limit as the limit on valid tokens, but I'm fine with trimming off `limit + 1`.